storage/s3.go

```diff
@@ -43,19 +42,18 @@ const (
 )
 
 // Re-used AWS sessions dramatically improve performance.
-var cachedSession func() *session.Session
+var cachedSessions func() *s3Session
 
 // Init creates a new global S3 session.
-func Init(opts Options) error {
-	sess, err := newSession(opts)
-	if err != nil {
-		return err
+func init() {
+	s3Sess := &s3Session{
```

Comment:
Why not just:

```go
func init() {
	sessionProvider = &s3Session{ // or maybe sessionFactory
		Mutex:    sync.Mutex{},
		sessions: map[Options]*session.Session{},
	}
}
```

and access it without function closure 
```go
awsSession, err := sessionProvider.newSession(ctx, opts) 
// rather than cachedSessions().newSession(ctx, opts)
```

Reply:
thanks. I had tried to mitigate potential harms of global variable by wrapping it in a function, since for example in [this test](https://github.com/peak/s5cmd/pull/194/files#diff-2c7a8270898140e540697cb13ca5406bR91) we have to modify the variable. Wrapping the variable with function might give slight warning when making such calls. But it might not be a big issue, so I changed according to your proposition.